Reference fields

SchrodingersGat opened the pull request:
This PR provides a major overhaul for how the "reference" fields are implemented:

- BuildOrder.reference
- SalesOrder.reference
- PurchasesOrder.reference

The *current* implementation provides a "prefix" and "regex" setting for each type of reference, e.g. for a BuildOrder the prefix might be `BO-` and the regex might be `(\d{4})`, meaning that the 'reference field' *must* be something like `BO-1234`

However this is overly complex, and does not allow us to implement the more complex behaviours as explored in https://github.com/inventree/InvenTree/discussions/3163

So this PR simultaneously provides an overhaul, simplification and expansion of functionality.

The "prefix" and "regex" concepts are consolidated into a configurable "reference pattern":

- Uses "pythonic" string formatting for named group replacement
- Allows simple pattern matching
- Allows specification of integer portion of reference

#### Example

Let's say the *Build Order Reference Pattern* is configured to be:

`BO-{ref:04d}`

![image](https://user-images.githubusercontent.com/10080325/177714592-baab85f6-1783-4051-bcdf-bf46de86b29e.png)

This pattern serves to:

1. Auto-generate the next build reference
2. Enforce a certain pattern for the build reference

Using python format strings allows us to expand this concept in the future, although the pattern substitution functionality is currently very limited.

---------------------------------
- Closes https://github.com/inventree/InvenTree/issues/3167
- Closes https://github.com/inventree/InvenTree/pull/3172

<a href="https://gitpod.io/#https://github.com/inventree/InvenTree/pull/3267"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

### TODO

- [x] Data migration for build order reference
- [x] Remove build order prefix setting
- [x] Update the "Sales Order" reference field accordingly
- [x] Update the "Purchase Order" reference field accordingly
- [x] Data migration for "Purchase Order" reference field
- [x] Data migration for "Sales Order" reference field
- [x] Remove purchase order prefix setting
- [x] Remove sales order prefix setting
- [x] Documentation - https://github.com/inventree/inventree-docs/pull/308
- [ ] Update unit tests for `inventree-python`
- [x] Perform migrations on demo dataset - https://github.com/inventree/demo-dataset/pull/26

### Future Extensions

This pattern idea can be also used for

- Serial numbers
- Batch codes
- Shipment codes
- Part name formatting

matmair commented:
@SchrodingersGat why not use f-string numerals? https://peps.python.org/pep-0498/ would make the syntax simpler and there are a lot of references. We could provide a context like this:
```
now: current timestamp
customer: current customer
...
serial: reference number
```

The sample string would be formatted like this:
`BO-{now:%Y.%W}-{serial:05d}`

Guide the f-strings https://realpython.com/python-string-formatting/#3-string-interpolation-f-strings-python-36

SchrodingersGat commented:
Using f-strings would be more powerful, but I feel less "accessible" to the general user, which is why I am moving away from user-facing regular expressions (per the discussion in https://github.com/inventree/InvenTree/pull/3172)

e.g. `PO-#####` vs `PO-{serial:05d}`

matmair commented:
@SchrodingersGat I'll let you finish before discussing it. Inventing a new pattern-matching system seems like unnecessary work to me.

SchrodingersGat commented:
@matmair I have updated the description now, but you'll probably want to have a read through the code to see what I'm going for.

Basically I have converted it to use pythonic string formatting, and trying to make a simple mixin class-based approach. LMK what you think. It will certainly need some more refinement.

matmair commented on InvenTree/InvenTree/models.py:
```diff
@@ -112,21 +112,170 @@ class ReferenceIndexingMixin(models.Model):
+
+    @classmethod
+    def generate_reference(cls):
+        """Generate the next 'refernce' field based on specified pattern"""
```
Typo

matmair commented:
@SchrodingersGat looks good, I added some non-blocking comments but should be good to go from my POV.

SchrodingersGat commented on InvenTree/InvenTree/serializers.py:
```diff
@@ -211,16 +208,6 @@ class Meta:
-
-    def validate_reference(self, value):
-        """Ensures the reference is not to big / small for the BigIntegerField."""
-        if extract_int(value) > models.BigIntegerField.MAX_BIGINT:
```
Has been added back in, thanks for pointing this out

matmair replied:
@SchrodingersGat we had that bug already, no need to refind the solution 😇

matmair commented:
@SchrodingersGat the settings for orders look pretty sparse - should we just merge them (SO, PO, BO) together?

miggland commented:
Hey guys, sorry for not reacting until now, just have been busy with other stuff.

This looks really good, a quick test confirms that this is what I was after in my original FR.
